### DISCUSSION TITLE
Need to run static code Analysis on particular block of code on an existing Apex class file

### DISCUSSION THREAD

**Discussion by meghachethu8223** [pmd/pmd#2915] [Q&A] [answered]
**Description:**

Hi Team,

I am using PMD to run static code analysis in Salesforce. I use Copado which is integrated with PMD and when I run static code analysis on an User story which has an old apex class, it runs SCA on complete file where I want to check only the violations of my changes not to scan entire class code as I will have to spend more time resolving on some one else code.

Please let me know if you have any feature to baseline and run SCA on only the new block change on an existing file.

Regards,
Meghana

**Comment by oowekyala** [comment, marked as answer]
Hi, there is no built-in mechanism to limit analysis to a single block of code. But depending on the report format you're using, you might be able to post-process the report to exclude violations based on their line numbers.

**Comment by meghachethu8223** [comment]
Thank you so much for the idea I will check with my team and see if we can try and implement something similar.
I saw a PMD document where it mentions "Incremental Analysis" feature to be enabled for similar kind of issues. If you have any idea on it please share.
Thanks!

**Comment by oowekyala** [comment]
Incremental analysis is a way to cache a report for files that haven't changed, but in the end the report we produce is the same (we only avoid reprocessing the file, and use the saved violations).

We also have a tool to produce the diff of two reports (sort of): https://github.com/pmd/pmd-regression-tester/
This was built to compare separate versions of pmd on the same codebase, so it doesn't have your feature yet. But it already has all the logic to produce the diff between two reports, so it could actually make sense to use this to instead compare the same version of pmd against slightly changed codebases. This would require some tweaks to the code I think

**Comment by meghachethu8223** [comment]
Yeah got it! Let me think on the process and see if I can get any similar solutions. Anyway thanks for sharing!

**Comment by rsoesemann** [comment]
Hi @meghachethu8223 I am not only a pmd maintainer but also just recently joined Copado. Feel free to reach out to me on rsoesemann@copado.com and we reach out to the team to see if and how we cover that.

**Comment by meghachethu8223** [comment]
Hi rsoesemann, Thank you so much!! I have checked your video regarding PMD its very informative and have connected with you in linkedin! I will contact you for any copado related queries. Thank you so much for letting me know.

**Comment by rsoesemann** [comment]
@meghachethu8223 I just confirmed with the team and the product manager that this indeed is not possible right now, but we fully agree that it is an improvement. I created a roadmap suggestion (lets see if they pick it) where for the PMD Settings you can say that it should only fail when X new issues are found in the CHANGED metadata. 

What do you think?

**Comment by meghachethu8223** [comment]
@rsoesemann Indeed I guess it will resolve my issue and that would be a great help.Thank you so much!

**Comment by meghachethu8223** [comment]
@rsoesemann Please keep me posted on the updates.

**Comment by rsoesemann** [comment]
Close this as this is not a PMD change request but a Copado one.
